### DISCUSSION TITLE
Keep-alive connections and invalid requests

### DISCUSSION THREAD

**Discussion by MSP-Greg** [puma/puma#3555] [Ideas]
Requests may be invalid due to being malformed or due to exceeding limits {either hard coded or Puma configuration options).

These requests take time and resources to parse, and may leave the parser in a state where it cannot differentiate between the end of the current request and the start of a subsequent request.

Hence, I’d like to move forward with changing Puma’s behavior such that any invalid request will result in a ‘4xx’ response status with a `connection: close` header and a closing of the socket.  The socket will not be ‘drained’.

Any objections or thoughts?

See #3540, #3552, along with ‘not yet finished’ work being done for the ‘long tail’ issue.

**Comment by dentarg** [comment]
> Hence, I’d like to move forward with changing Puma’s behavior such that any invalid request will result in a ‘4xx’ response status with a `connection: close` header and a closing of the socket. The socket will not be ‘drained’.

Sounds good to me.

For the record, what is the behaviour today?

**Reply by MSP-Greg** [reply]
I believe current behavior does generate the '4xx', but it may not close the socket.

There is a generic issue, and that is once a request is considered invalid, can a parser read thru the rest of the request characters and determine where the request ends? In many cases, that probably isn't possible, and it may waste CPU time.

The client maintains a buffer of what's been read but not parsed, and parsing stops on invalid data.  So, for an invalid request, the buffer may contain garbage.

Now, assume this is a 'keep-alive' connection.  So, as Puma is currently set up, it will check the buffer and/or read the socket.  If that data contains the remains of the invalid request, it will look like garbage (certainly not a request start line).  Often, Puma will throw the 'is this an SSL connection on a TCP socket' error.

Which leads to the conclusion that Puma has to close the socket when an invalid request is received.
